File: crates/avenger-validate/cel/parity.mjs
import {readFileSync} from 'node:fs'
import {Bundle} from './avenger_validate_cel.mjs'

const b = new Bundle(JSON.parse(readFileSync(process.argv[2], 'utf8')))
const rows = readFileSync(process.argv[3], 'utf8').trim().split('\n').map(l => JSON.parse(l))
const key = is => JSON.stringify(is.map(i => [i.layer, i.step]))
const t = performance.now()
const got = rows.map(r => key(b.validate(r.steps)))
const us = ((performance.now() - t) / rows.length) * 1000
const t2 = performance.now()
for (let k = 0; k < 20; k++) for (const r of rows) b.validate(r.steps)
const warm = ((performance.now() - t2) / rows.length / 20) * 1000
const bad = rows.filter((r, k) => key(r.issues) !== got[k])
console.log(`cel-js: ${rows.length - bad.length} of ${rows.length} agree with Rust (${us.toFixed(0)} µs a pipeline first pass, ${warm.toFixed(1)} µs warm)`)
for (const r of bad.slice(0, 5)) console.log('  differs:', JSON.stringify(r.steps), key(r.issues))

File: crates/avenger-validate/cel/parity.py
"""Check that cel-python and cel-js give what the Rust validator gives, on
every pipeline of experiment 7: `(layer, step)` of each layer 1-2 issue.

    avenger-validate export-cel > /tmp/bundle.json
    avenger-validate corpus <results dir> --jsonl > /tmp/corpus.jsonl
    python parity.py /tmp/bundle.json /tmp/corpus.jsonl
"""

import json
import subprocess
import sys
import time
from pathlib import Path

from avenger_validate_cel import Bundle

bundle_path, corpus_path = sys.argv[1], sys.argv[2]
rows = [json.loads(line) for line in open(corpus_path)]
want = [[(i["layer"], i["step"]) for i in r["issues"]] for r in rows]

b = Bundle(json.load(open(bundle_path)))
t = time.perf_counter()
got = [[(i["layer"], i["step"]) for i in b.validate(r["steps"])] for r in rows]
py_us = (time.perf_counter() - t) / len(rows) * 1e6
bad = [k for k, (w, g) in enumerate(zip(want, got)) if w != g]
t = time.perf_counter()
for r in rows:
    b.validate(r["steps"])
warm_us = (time.perf_counter() - t) / len(rows) * 1e6
print(f"cel-python: {len(rows) - len(bad)} of {len(rows)} agree with Rust ({py_us:.0f} µs a pipeline first pass, {warm_us:.0f} µs warm)")
for k in bad[:5]:
    print("  differs:", rows[k]["steps"], want[k], got[k])

js = Path(__file__).with_name("parity.mjs")
out = subprocess.run(["node", str(js), bundle_path, corpus_path], capture_output=True, text=True, check=True).stdout
print(out.strip())
